Send Logstash Metrics via OTLP (#18857)
elastic/logstash: docs/reference/logstash-settings-file.md

| `allow_superuser` | Setting to `true` to allow or `false` to block running Logstash as a superuser. | `false` |
| `pipeline.buffer.type` | Determine where to allocate memory buffers, for plugins that leverage them.Defaults to `heap` but can be switched to `direct` to instruct Logstash to prefer allocation of buffers in direct memory. | `heap` Check out [Buffer Allocation types](/reference/jvm-settings.md#off-heap-buffers-allocation) for more info. |
| `otel.metrics.enabled` | Enable or disable OpenTelemetry metrics export. See [Monitoring with OpenTelemetry](/reference/monitoring-with-opentelemetry.md). | `false` |
| `otel.metrics.endpoint` | The OTLP endpoint URL for metrics export. For gRPC, typically port 4317. For HTTP, typically port 4318. | `http://localhost:4317` |
| `otel.metrics.interval` | Export interval in seconds. Controls how frequently metrics are sent to the OTLP endpoint. | `10` |
| `otel.metrics.protocol` | Protocol to use for OTLP export. Valid values are `grpc` or `http`. | `grpc` |

> | `otel.metrics.authorization_header` | Authorization header for authenticated OTLP endpoints. Examples: `ApiKey xxx` or `Bearer xxx`. | *N/A* |

Author: Hey Rye, thanks for suggesting this. Where did you see that only the`OtlpGrpcMetricExporterBuilder` has a way of providing ssl cert and key? I'm seeing that both http and grpc metric exporter builders support it.
[http](https://github.com/open-telemetry/opentelemetry-java/blob/6336abdf76497335287e96fc5cd18566b4594930/exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/http/metrics/OtlpHttpMetricExporterBuilder.java#L172-L191)
[grpc](https://github.com/open-telemetry/opentelemetry-java/blob/6336abdf76497335287e96fc5cd18566b4594930/exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/metrics/OtlpGrpcMetricExporterBuilder.java#L180-L199)

> | `otel.metrics.authorization_header` | Authorization header for authenticated OTLP endpoints. Examples: `ApiKey xxx` or `Bearer xxx`. | *N/A* |

Author: I've added support for a ca, client cert and client key in this commit
https://github.com/elastic/logstash/pull/18857/changes/063698464989b7fcca15e9f6a23888ec5bae1e2e
It's added for both http and grpc protocols

> | `otel.metrics.authorization_header` | Authorization header for authenticated OTLP endpoints. Examples: `ApiKey xxx` or `Bearer xxx`. | *N/A* |

Reviewer: I must have been looking at out-of-date docs. Food to see that both protocols support identity key material.

| `allow_superuser` | Setting to `true` to allow or `false` to block running Logstash as a superuser. | `false` |
| `pipeline.buffer.type` | Determine where to allocate memory buffers, for plugins that leverage them.Defaults to `heap` but can be switched to `direct` to instruct Logstash to prefer allocation of buffers in direct memory. | `heap` Check out [Buffer Allocation types](/reference/jvm-settings.md#off-heap-buffers-allocation) for more info. |
| `otel.metrics.enabled` | Enable or disable OpenTelemetry metrics export. See [Monitoring with OpenTelemetry](/reference/monitoring-with-opentelemetry.md). | `false` |
| `otel.metrics.endpoint` | The OTLP endpoint URL for metrics export. For gRPC, typically port 4317. For HTTP, typically port 4318. | `http://localhost:4317` |
| `otel.metrics.interval` | Export interval with time unit (e.g., `"10s"`, `"1m"`). Controls how frequently metrics are sent to the OTLP endpoint. | `"10s"` |
| `otel.metrics.protocol` | Protocol to use for OTLP export. Valid values are `grpc` or `http`. | `grpc` |
| `otel.metrics.authorization_header` | Authorization header for authenticated OTLP endpoints. Examples: `ApiKey xxx` or `Bearer xxx`. | *N/A* |

> | `otel.resource.attributes` | Additional OpenTelemetry resource attributes as comma-separated key=value pairs. Example: `environment=production,cluster=us-west`. | *N/A* |

Reviewer 2: Currently, users who send Logstash metrics to Elasticsearch via OTLP will find their data in `metrics-generic.otel-default` unless they manually set `otel.resource.attributes: "data_stream.dataset=logstash"` in `logstash.yml` — which requires knowing an Elasticsearch-specific OTel convention that isn't obvious.

WDYT about making a new setting `otel.metrics.dataset` setting (defaulting to `"logstash"`) that gets applied as the `data_stream.dataset` resource attribute internally in `OtelMetricsService.java`, before user-provided `otel.resource.attributes` are merged in (so an explicit `data_stream.dataset=<custom>` in `otel.resource.attributes` still wins).

Would be IMHO a stronger UX and simpler for users.

Reviewer 2: Also, when trying to work with the data in Kibana it's a bit awkward as for some reason `TS metrics-logstash.otel-default` doesn't show the nice charts like other streams do.

Any thoughts on next steps for completing the story here? Maybe with an integration etc? Doesn't need to be in this PR but I'm wondering if there are any considerations for Kibana users that we should be thinking about now beyond what I articulated above.

Author: @perk what do you think about the setting that Andrew suggested?

Reviewer 3: Hey @estolfo @andrewvc I very much agree with the suggestion. The data comes from Logstash, so it should land in the default Logstash monitoring dataset rather than generic OTLP one.

Author: I've added support for the data_stream.dataset resource attribute in this commit:
https://github.com/elastic/logstash/pull/18857/changes/10f7f647b9bf7cd2c81de63784acebe9572080b7

Author: hey @perk and @andrewvc 
in working on building the integration for these metrics, I've learned that the data_stream.dataset value must exactly match what the integration expects. If the user sets it to something else, the integration won't pick up the data. So what do you think-- is it worth letting the user set this value themselves and rely on documentation to ensure that the user knows it should be "logstash" (or `logstash.otel`, see below) for the integration to work?

Also note that the otel_collector_internal_telemetry integration package uses the convention `<product>.otel` as the dataset name ([collectortelemetry.otel](https://github.com/elastic/integrations/blob/d6f3eb25addabacc3a014cae6c9d829e118bfcf6/packages/otel_collector_internal_telemetry/manifest.yml#L20)). With that said, the Logstash dataset should be probably `logstash.otel`, not just `logstash`.

Reviewer 2: I agree that it should be `logstash.otel`. It's a bit ugly, but I don't see any better alternative.

Author: as it turns out, I discovered while testing today that when data is sent to the motel endpoint, it automatically appends `.otel` to the dataset name for OTLP traffic. So we should leave the dataset name as `logstash` because it will become `logstash.otel` anyway.

Reviewer 2: Hmmm, what about for self-managed users using this with the ES OTLP endpoint?

| `allow_superuser` | Setting to `true` to allow or `false` to block running Logstash as a superuser. | `false` |
| `pipeline.buffer.type` | Determine where to allocate memory buffers, for plugins that leverage them.Defaults to `heap` but can be switched to `direct` to instruct Logstash to prefer allocation of buffers in direct memory. | `heap` Check out [Buffer Allocation types](/reference/jvm-settings.md#off-heap-buffers-allocation) for more info. |
| `otel.metrics.enabled` | Enable or disable OpenTelemetry metrics export. See [Monitoring with OpenTelemetry](/reference/monitoring-with-opentelemetry.md). | `false` |
| `otel.exporter.otlp.metrics.endpoint` | OTLP metrics endpoint URL. For gRPC, typically port 4317. For HTTP, typically port 4318. | `http://localhost:4317` |
| `otel.metric.export.interval` | Export interval with time unit (e.g., `"10s"`, `"1m"`). Controls how frequently metrics are sent to the OTLP endpoint. | `"10s"` |

> | `otel.exporter.otlp.metrics.protocol` | Protocol to use for OTLP export. Valid values are `grpc` or `http`. | `grpc` |

Reviewer 2: Same concerns as previously WRT these new confusing prefixes

| `path.plugins` | Where to find custom plugins. You can specify this setting multiple times to include  multiple paths. Plugins are expected to be in a specific directory hierarchy:  `PATH/logstash/TYPE/NAME.rb` where `TYPE` is `inputs`, `filters`, `outputs`, or `codecs`,  and `NAME` is the name of the plugin. | Platform-specific. See [Logstash Directory Layout](/reference/dir-layout.md). |
| `allow_superuser` | Setting to `true` to allow or `false` to block running Logstash as a superuser. | `false` |
| `pipeline.buffer.type` | Determine where to allocate memory buffers, for plugins that leverage them.Defaults to `heap` but can be switched to `direct` to instruct Logstash to prefer allocation of buffers in direct memory. | `heap` Check out [Buffer Allocation types](/reference/jvm-settings.md#off-heap-buffers-allocation) for more info. |

> | `otel.metrics.enabled` | Enable or disable OpenTelemetry metrics export. See [Monitoring with OpenTelemetry](/reference/monitoring-with-opentelemetry.md). | `false` |

Reviewer 4: ```suggestion
| `otel.metrics.enabled` {applies_to}`stack: preview 9.5.0+` | Turn on or off OpenTelemetry metrics export. Refer to [Monitoring with OpenTelemetry](/reference/monitoring-with-opentelemetry.md). | `false` |
```

> | `otel.exporter.otlp.endpoint` | OTLP metrics endpoint URL. For gRPC, typically port 4317. For HTTP, typically port 4318. When using HTTP protocol, `/v1/metrics` is automatically appended if not already present, and the default port (80 for `http://`, 443 for `https://`) is added if no port is specified. | `http://localhost:4317` |

Reviewer 4: ```suggestion
| `otel.exporter.otlp.endpoint` {applies_to}`stack: preview 9.5.0+` | OTLP metrics endpoint URL. For gRPC, typically port 4317. For HTTP, typically port 4318. When using HTTP protocol, `/v1/metrics` is automatically appended if not already present, and the default port (80 for `http://`, 443 for `https://`) is added if no port is specified. | `http://localhost:4317` |
```

> | `otel.metric.export.interval` | Export interval with time unit (e.g., `"10s"`, `"1m"`). Controls how frequently metrics are sent to the OTLP endpoint. | `"10s"` |

Reviewer 4: ```suggestion
| `otel.metric.export.interval` {applies_to}`stack: preview 9.5.0+` | Export interval with time unit. Controls how frequently metrics are sent to the OTLP endpoint. For example, `"10s"`, `"1m"`. | `"10s"` |
```

| `otel.resource.attributes` | Additional OpenTelemetry resource attributes as comma-separated key=value pairs. Example: `environment=production,cluster=us-west`. | *N/A* |
| `otel.service.name` | Service name for metrics. Can also be set via `otel.service.name` system property. | `logstash` |
| `otel.dataset` | Value for the `data_stream.dataset` resource attribute. When sending to the Elastic ingest endpoint, `.otel` is automatically appended — the default `logstash` produces `data_stream.dataset: logstash.otel` and index `metrics-logstash.otel-<namespace>`. This setting is only configurable via `logstash.yml` and has no corresponding system property. | `logstash` |
| `otel.exporter.otlp.certificate` | Path to a PEM-encoded trusted CA certificate for verifying the OTLP endpoint's TLS certificate. Required when the endpoint uses a self-signed or private CA. | *N/A* |
| `otel.exporter.otlp.client.key` | Path to a PEM-encoded client private key for mutual TLS (mTLS). Must be set together with `otel.exporter.otlp.client.certificate`. | *N/A* |

> | `otel.exporter.otlp.client.certificate` | Path to a PEM-encoded client certificate for mutual TLS (mTLS). Must be set together with `otel.exporter.otlp.client.key`. | *N/A* |

Reviewer 4: ```suggestion
| `otel.exporter.otlp.protocol` {applies_to}`stack: preview 9.5.0+` | Protocol to use for OTLP export. Valid values are `grpc` or `http`. | `grpc` |
| `otel.exporter.otlp.headers` {applies_to}`stack: preview 9.5.0+` | HTTP headers as comma-separated `key=value` pairs to include in every OTLP request. Example: `Authorization=ApiKey xxx` or `Authorization=Bearer xxx,X-Custom=foo`. | *N/A* |
| `otel.resource.attributes` {applies_to}`stack: preview 9.5.0+` | Additional OpenTelemetry resource attributes as comma-separated key=value pairs. Example: `environment=production,cluster=us-west`. | *N/A* |
| `otel.service.name` {applies_to}`stack: preview 9.5.0+` | Service name for metrics. Can also be set via `otel.service.name` system property. | `logstash` |
| `otel.dataset` {applies_to}`stack: preview 9.5.0+` | Value for the `data_stream.dataset` resource attribute. When sending to the Elastic ingest endpoint, `.otel` is automatically appended — the default `logstash` produces `data_stream.dataset: logstash.otel` and index `metrics-logstash.otel-<namespace>`. This setting is only configurable via `logstash.yml` and has no corresponding system property. | `logstash` |
| `otel.exporter.otlp.certificate` {applies_to}`stack: preview 9.5.0+` | Path to a PEM-encoded trusted CA certificate for verifying the OTLP endpoint's TLS certificate. Required when the endpoint uses a self-signed or private CA. | *N/A* |
| `otel.exporter.otlp.client.key` {applies_to}`stack: preview 9.5.0+` | Path to a PEM-encoded client private key for mutual TLS (mTLS). Must be set together with `otel.exporter.otlp.client.certificate`. | *N/A* |
| `otel.exporter.otlp.client.certificate` {applies_to}`stack: preview 9.5.0+` | Path to a PEM-encoded client certificate for mutual TLS (mTLS). Must be set together with `otel.exporter.otlp.client.key`. | *N/A* |
```
